src/VisualStudio/Core/Def/Implementation/Workspace/SourceGeneratedFileManager.cs

```diff
@@ -0,0 +1,422 @@
+            _temporaryDirectory = Path.Combine(Path.GetTempPath(), "VisualStudioSourceGeneratedDocuments");
+            _visualStudioWorkspace = visualStudioWorkspace;
+
+            Directory.CreateDirectory(_temporaryDirectory);
+
+            _runningDocumentTable = _serviceProvider.GetService<SVsRunningDocumentTable, IVsRunningDocumentTable>();
+            _runningDocumentTableEventTracker = new RunningDocumentTableEventTracker(
+                threadingContext,
+                editorAdaptersFactoryService,
+                _runningDocumentTable,
+                this);
+        }
+
+        public void NavigateToSourceGeneratedFile(Project project, ISourceGenerator generator, string generatedSourceHintName, TextSpan sourceSpan)
+        {
+            // We will create an file name to represent this generated file; the Visual Studio shell APIs imply you can use a URI,
+            // but most URIs are blocked other than file:// and http://; they also get extra handling to attempt to download the file so
+            // those aren't really usable anyways.
+
+            // The file name we generate here is chosen to match the compiler's choice, so the debugger can recognize the files should match.
+            // We use the ProjectId as a parent directory so we can extract that out later too.
+            var generatorType = generator.GetType();
+            var temporaryFilePath = Path.Combine(_temporaryDirectory, $"{project.Id.Id}\\{generatorType.Module.ModuleVersionId}_{generatorType.FullName}_{generatedSourceHintName}");
+            Directory.CreateDirectory(Path.GetDirectoryName(temporaryFilePath));
+            File.WriteAllText(temporaryFilePath, "");
+
+            var openDocumentService = _serviceProvider.GetService<SVsUIShellOpenDocument, IVsUIShellOpenDocument>();
+            var hr = openDocumentService.OpenDocumentViaProject(
+                temporaryFilePath,
+                VSConstants.LOGVIEWID.TextView_guid,
+                out _,
+                out _,
+                out _,
+                out var windowFrame);
+
+            if (ErrorHandler.Succeeded(hr) && windowFrame != null)
+            {
+                windowFrame.Show();
+            }
+        }
+
+        public bool TryParseGeneratedFilePath(
+            string filePath,
+            [NotNullWhen(true)] out ProjectId? projectId,
+            [NotNullWhen(true)] out string? generatorTypeName,
+            [NotNullWhen(true)] out string? generatedSourceHintName)
+        {
+            if (!filePath.StartsWith(_temporaryDirectory))
+            {
+                projectId = null;
+                generatorTypeName = null;
+                generatedSourceHintName = null;
+                return false;
+            }
+
+            string fileName = Path.GetFileName(filePath);
+            string[] parts = fileName.Split(new[] { '_' }, count: 3);
+
+            generatorTypeName = parts[1];
```

Comment:
❔ What if the generator name has a `_` in it?

Reply:
This is being discussed in https://github.com/dotnet/roslyn/pull/46949#discussion_r476921385.